torchgeo/models/farseg.py

```diff
@@ -0,0 +1,227 @@
+# Copyright (c) Microsoft Corporation. All rights reserved.
+# Licensed under the MIT License.
+
+"""Foreground-Aware Relation Network (FarSeg) implementations."""
+
+import torch
+import torch.nn as nn
+import torch.nn.functional as F
+from torchvision.models import resnet
+from torchvision.ops import FeaturePyramidNetwork as FPN
+from typing import List, Tuple
+import math
+from collections import OrderedDict
+
+
+class FarSeg(nn.Module):
+    """Foreground-Aware Relation Network (FarSeg).
+
+    'Foreground-Aware Relation Network for Geospatial Object Segmentation
+    in High Spatial Resolution Remote Sensing Imagery', Zheng et al. (2020)
+    If you use this model in your research, please cite the following paper:
+
+    * https://arxiv.org/pdf/2011.09766.pdf
+    @inproceedings{zheng2020foreground,
+      title={Foreground-Aware Relation Network for Geospatial Object Segmentation in High Spatial Resolution Remote Sensing Imagery},
+      author={Zheng, Zhuo and Zhong, Yanfei and Wang, Junjue and Ma, Ailong},
+      booktitle={Proceedings of the IEEE/CVF Conference on Computer Vision and Pattern Recognition},
+      pages={4096--4105},
+      year={2020}
+    }
+    """
+
+    def __init__(self,
+                 in_channels: int,
+                 backbone: str,
+                 classes: int,
+                 backbone_pretrained: bool = True
+                 ) -> None:
+        """Initializes the FarSeg.
+
+        Args:
+            in_channels: number of channels per input image
+            backbone: name of ResNet backbone
+            classes: number of output segmentation classes
+            backbone_pretrained: whether to use pretrained weight for backbone
+                (default=True)
+        """
+        super().__init__()
+        if backbone in ['resnet18', 'resnet34']:
+            max_channels = 512
+        elif backbone in ['resnet50', 'resnet101']:
+            max_channels = 2048
+        else:
+            raise ValueError(f'unknown backbone: {backbone}.')
+        self.backbone = getattr(resnet, backbone)(pretrained=backbone_pretrained)
+
+        self.reset_in_channels_for_backbone(in_channels)
+
+        self.fpn = FPN(in_channels_list=[max_channels // (2 ** (3 - i)) for i in range(4)], out_channels=256)
+        self.fsr = FSRelation(max_channels, [256] * 4, 256, True)
+        self.decoder = LightWeightDecoder(256, 128, classes)
+
+    def reset_in_channels_for_backbone(self, in_channels):
+        if in_channels == 3:
+            return
+
+        self.backbone.add_module('conv1',
+                                 nn.Conv2d(in_channels,
+                                           64, kernel_size=7, stride=2, padding=3,
+                                           bias=False))
+
+    def backbone_forward(self, x: torch.Tensor) -> Tuple[torch.Tensor]:
```

Comment:
All of these functions will need docstrings explaining what they do, what their input arguments are, and what their return values are.